### PR TITLE
Scripting Support

### PR DESCRIPTION
**Pull request by zicklag**
Hey there :wave: 

Here's a PR that attempts to migrate from using `TypeId` to using `StorageId` in an effort to start to address #96.

I may have missed a spot here or there, and this doesn't address anything an API change to facilitate creating custom components yet ( necessarily ) it just switches everything I could find that should take a `StorageId` instead of a `TypeId`.

Definitely needs some review, but all of the tests pass.

### CODE REVIEW

File: src/storage/all/mod.rs [hunk 1]
```diff
@@ -28,7 +28,8 @@ use parking_lot::{lock_api::RawRwLock as _, RawRwLock};
 // we use a HashMap, it can reallocate, but even in this case the storages won't move since they are boxed
 pub struct AllStorages {
     lock: RawRwLock,
-    storages: UnsafeCell<HashMap<TypeId, Storage, BuildHasherDefault<TypeIdHasher>>>,
+    // TODO: Re-include the TypeIdHasher as a StorageIdHasher
+    storages: UnsafeCell<HashMap<StorageId, Storage /*, BuildHasherDefault<TypeIdHasher>*/>>,
```
Comment (zicklag) [inline src/storage/all/mod.rs:32]: Do we need to re-implement our own hash now that we are using `StorageId` now, or is the default one going to be fine?

Comment (leudz) [reply src/storage/all/mod.rs:32]: The default one is (and always was) fine. It was a good opportunity to learn about `Hash` and `Hasher` but it's probably a good time to remove the custom hasher, it'll simplify things.

**Comment by leudz** [comment]
I think we have to discuss the api before I can merge this PR. Based on the api, if some part can't be used with ffi views, it'll make more sense to use `TypeId` in this part of the codebase.

I think a good example of this is systems, they can't work without types. I think ffi systems will do the same thing as workloads.

**Comment by zicklag** [comment]
That makes sense. I've got a plan for a prototype API for scripted systems and components in mind so I'll rename this PR to "Scripting Support" and get those changes up when I have them so we can discuss it.

I'm not sure if I'm missing something huge or not, but it looks like it won't be too difficult. So far the code-base has made complete sense to me which is great. I wasn't sure if I would be completely lost if I looked into it.

Anyway we'll see how well that goes. :smiley: :+1:

**Comment by leudz** [comment]
I'm closing it but I'll take another look inside when I finally try to do scripting :sweat_smile:

**Pull request closed without merging**